File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.java
import java.util.*;

public class PreferentialAttachment {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public double preferentialAttachmentScore(int u, int v) {
        if (!adjacency.containsKey(u) || !adjacency.containsKey(v)) {
            return 0.0;
        }
        return (double) adjacency.get(u).size() * adjacency.get(v).size();
    }

    public List<Map.Entry<int[], Double>> scoreAllPairs() {
        List<Map.Entry<int[], Double>> scores = new ArrayList<>();
        List<Integer> nodeList = new ArrayList<>(nodes);

        for (int i = 0; i < nodeList.size(); i++) {
            for (int j = i + 1; j < nodeList.size(); j++) {
                int u = nodeList.get(i);
                int v = nodeList.get(j);
                double score = preferentialAttachmentScore(u, v);
                if (score > 0) {
                    scores.add(new AbstractMap.SimpleEntry<>(new int[]{u, v}, score));
                }
            }
        }

        scores.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        return scores;
    }

    public List<Map.Entry<int[], Double>> topKPredictions(int k) {
        List<Map.Entry<int[], Double>> scores = scoreAllPairs();
        return scores.subList(0, Math.min(k, scores.size()));
    }
}

File: frontend/src/components/benchmark/content/preferential_attachment/implementations/preferential_attachment.go
package main

import (
    "sort"
)

type PreferentialAttachment struct {
    adjacency map[int]map[int]bool
    nodes     map[int]bool
}

func NewPreferentialAttachment() *PreferentialAttachment {
    return &PreferentialAttachment{
        adjacency: make(map[int]map[int]bool),
        nodes:     make(map[int]bool),
    }
}

func (pa *PreferentialAttachment) AddEdge(u, v int) {
    if pa.adjacency[u] == nil {
        pa.adjacency[u] = make(map[int]bool)
    }
    if pa.adjacency[v] == nil {
        pa.adjacency[v] = make(map[int]bool)
    }
    pa.adjacency[u][v] = true
    pa.adjacency[v][u] = true
    pa.nodes[u] = true
    pa.nodes[v] = true
}

func (pa *PreferentialAttachment) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        pa.AddEdge(edge[0], edge[1])
    }
}

func (pa *PreferentialAttachment) PreferentialAttachmentScore(u, v int) float64 {
    if pa.adjacency[u] == nil || pa.adjacency[v] == nil {
        return 0.0
    }
    return float64(len(pa.adjacency[u]) * len(pa.adjacency[v]))
}

func (pa *PreferentialAttachment) ScoreAllPairs() [][3]interface{} {
    var scores [][3]interface{}
    var nodeList []int
    for node := range pa.nodes {
        nodeList = append(nodeList, node)
    }

    for i := 0; i < len(nodeList); i++ {
        for j := i + 1; j < len(nodeList); j++ {
            u := nodeList[i]
            v := nodeList[j]
            score := pa.PreferentialAttachmentScore(u, v)
            if score > 0 {
                scores = append(scores, [3]interface{}{u, v, score})
            }
        }
    }

    sort.Slice(scores, func(i, j int) bool {
        return scores[i][2].(float64) > scores[j][2].(float64)
    })

    return scores
}

func (pa *PreferentialAttachment) TopKPredictions(k int) [][3]interface{} {
    scores := pa.ScoreAllPairs()
    if k > len(scores) {
        k = len(scores)
    }
    return scores[:k]
}